src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs

```diff
@@ -1349,32 +1348,181 @@ private LazyStringValue TryGetTimedCounterTag(JsonOperationContext context, stri
+            {
+                if (_next == null)
+                {
+                    if (ToAppend.MoveNext() == false)
+                    {
+                        if (_current != null) 
+                            Last = _current.Timestamp;
+
+                        _current = null;
+                        return false;
+                    }
+
+                    var time = EnsureMillisecondsPrecision(ToAppend.Current!.Timestamp);
+                    if (_current == null) 
+                        First = time;
+
+                    _current ??= new SingleResult();
+
+                    ToAppend.Current.CopyTo(_current);
+                    _current!.Timestamp = time;
+                    AssertResult(_current);
+                }
+                else
+                {
+                    _current = _next;
+                    _next = null;
+                }
+
+                while (true)
+                {
+                    if (ToAppend.MoveNext() == false)
+                    {
+                        _next = null;
+                        break;
+                    }
+
+                    _next ??= new SingleResult();
+
+                    ToAppend.Current.CopyTo(_next);
+                    _next!.Timestamp = EnsureMillisecondsPrecision(_next.Timestamp);
+
+                    AssertResult(_next);
+
+                    if (_current.Timestamp > _next.Timestamp)
+                        throw new InvalidDataException($"The entries of '{Name}' incremental time-series for document '{DocumentId}' must be sorted by their timestamps. " +
+                                                       $"Got: current '{_current.Timestamp:O}'.");
+
+                    if (_current.Timestamp == _next.Timestamp)
+                    {
+                        var tagCompare = _current.Tag.CompareTo(_next.Tag);
+                        if (tagCompare > 0)
+                            throw new InvalidDataException(
+                                $"Entries to append are out of tag order for '{Name}' time-series for document '{DocumentId}' at '{_current.Timestamp:O}' (current:{_current.Tag}, next:{_next.Tag})");
+
+                        // note that we can have a situation like Inc(1,-1) followed by Inc(1,1) at the same timestamp
+                        // which will cause out of order that we don't what to handle and will throw
+                        if (tagCompare == 0)
+                        {
+
+                            if (FromReplication)
+                                throw new InvalidDataException(
+                                    $"Got an invalid segment from replication for '{Name}' time-series for document '{DocumentId}', segment contain duplicate timestamp and tag at '{_current.Timestamp:O}'");
+
+                            for (int i = 0; i < _next.Values.Length; i++)
+                            {
+                                _current.Values.Span[i] += _next.Values.Span[i];
```

Comment:
Don't we need to worry about the difference in the lengths of those two arrays?

Reply:
This code path is for entries we get from the user, so in the client we validate it has the same length.
But this indeed might be an issue we we will use REST... at this point I will simply throw